### DISCUSSION TITLE
How can a plugin override the language for a specific filename?

### DISCUSSION THREAD

**Discussion by tobysmith568** [prettier/prettier#16621] [Q&A] [answered]
Hey,

Is it possible for a plugin to override the language for a preconfigured list of file names, even when their file extension matches a built-in language?

For example, if a plugin adds the new language `SomethingScript` like so, can it apply this new language & parser specifically to files called `something.js`?

```ts
export const languages: Partial<SupportLanguage>[] = [
  {
    name: "SomethingScript",
    parsers: ["something-script"]
  }
];

export const parsers: Record<string, Parser> = {
  "something-script": {
    astFormat: "something-script-ast"
    // TODO
  }
};

export const printers: Record<string, Printer> = {
  "something-script-ast": {
    // TODO
  }
};
```

I've seen the [`filenames`](https://prettier.io/docs/en/api#prettiergetsupportinfo:~:text=%5B%5D%3B-,filenames,-%3F%3A%20string) field on the `SupportLanguage` type that is used by the `languages` export, however, I've been having no luck with the following:
```ts
export const languages: Partial<SupportLanguage>[] = [
  {
    name: "SomethingScript",
    parsers: ["something-script"],
    filenames: ["something.js"]
  }
];
```

Thanks in advance!

**Comment by arham766** [comment, marked as answer]
Yes, `filenames` is the right mechanism, and it does beat extension inference. When Prettier infers a language for a file it matches in two passes: first it checks every language's `filenames` for an exact basename match, and only if nothing matched does it fall back to matching `extensions`. So a plugin language with `filenames: ["something.js"]` wins over the built-in JavaScript language for a file named `something.js`, even though `.js` would otherwise match babel. Plugin languages also take priority over the core ones (the search list is built with the plugins reversed, so later-loaded plugins win over earlier ones and over the defaults).

A few things to check if it's not kicking in:

- The match is against the file's basename, exactly. `filenames: ["something.js"]` matches `src/something.js` but not `something.test.js` — no globs.
- Your parser needs to be minimally complete (`parse`, `astFormat`, `locStart`, `locEnd`) and the printer registered under that `astFormat`. If the parser throws, it can look like the language was never picked.
- You can verify what Prettier decided with `prettier --file-info something.js --plugin ./your-plugin.js` — it prints the inferred parser without formatting anything.

Also worth knowing: on newer Prettier versions a language can define an `isSupported({ filepath })` callback, which is checked after filenames/extensions and lets you match arbitrary paths when a fixed filename list isn't flexible enough.

**Reply by fisker** [reply]
I'm impressed. Is it answered by AI? Thanks anyway @arham766!

**Reply by fisker** [reply]
BTW, there might be bugs in old versions, but the current version (3.9 as writing) checks plugins from last to first.
